executor/executor_linux.h

```diff
@@ -259,3 +262,21 @@ static void setup_machine()
 	if (!write_file("/sys/kernel/debug/x86/nmi_longest_ns", "10000000000"))
 		printf("write to /sys/kernel/debug/x86/nmi_longest_ns failed: %s\n", strerror(errno));
 }
+
+static void df_open(df_detection* dfetch)
+{
+	int fd = open("/sys/kernel/debug/df_detection", O_RDWR);
+	if (fd == -1)
+		fail("open of /sys/kernel/debug/df_detection failed");
+	if (dup2(fd, dfetch->fd) < 0)
```

Comment:
I'm not really sure what is going on here. In most cases dfetch->fd is initially zero, passing it by value to dup2() will probably just close fd 0 and dup the new descriptor into it. Because many programs are going to continue using fd 0, that'll be a mess.
I think once you've checked that `fd != -1` you can just assign `fd` to `dfetch->fd`.

Reply:
As discussed yesterday, this code actually makes sense in the case where dfetch->fd is initialized with a predefined constant, like this is done in executor.cc

On the other hand, we don't really need to keep this descriptor around. Instead, we could make `df_enable()` a function that opens the file, calls the ioctl and closes the file.